The big cleanup for ThingTalk 2.0

gcampax opened the pull request:
This PR contains a large cleanup of genienlp, removing everything that is obsolete.

The major change is that masked tokenization is no more, because ThingTalk 2.0 doesn't use weird prefixes for parameter names or enums. Instead, the few remaining non-word tokens in ThingTalk (entities, device names) are added as special tokens to the Huggingface tokenizers, and everything else is subword tokenized. This has two practical effects:
- we greatly reduce the number of untrained embeddings
- tokenization works with any Huggingface model without code changes

GloVe and the MQAN encoder are gone. As a result, the numericalization code is cleaned up.
The internal abstraction of Seq2Seq with separate embeddings, encoder, and decoder was getting very messy, so that's gone too. There are now two models: BertLSTM and Bart.
Each model uses `--pretrained_model` to know which Huggingface model to base itself from. BertLSTM works with any Huggingface model. In principle, Bart could be T5 as well, but Huggingface doesn't have "AutoModelForConditionalGeneration" (or at least I didn't find it).

The command-line arguments are also cleaned up, which breaks backward compatibility. I think it's okay because ThingTalk broke compatibility anyway so old models don't work.
The plan is to make a 0.5.0 release before this is merged (for the sake of Schema2QA and AutoQA), then merge this, and rerun any outstanding paper and product work.

This is marked as draft because tests don't pass yet, and also because I want your opinion soon.

gcampax committed:
Remove GloVe embeddings

Not useful in a modern NLP pipeline

gcampax committed:
Remove custom masked tokenizer classes

Going forward for ThingTalk 2.0, we'll apply subword tokenization
to all tokens except to device names and entity names (which
have long namespace prefixes). Because we only have a fixed few
such tokens to worry about, we can add them to the Huggingface
tokenizers as special tokens, and remove our custom tokenizer
classes. This means we can use T5 or other newer models without
code changes.

gcampax committed:
Remove the option to freeze encoder embeddings

The option does not make sense if the whole encoder is just
"embedding" (that is, BERT)

gcampax committed:
Remove all unused MQAN-like encoders

Only leave the "Identity" encoder (aka the BERT encoder), and
remove the command-line options, which are now redundant.

gcampax committed:
Remove --pretrain_context option

Unused

gcampax committed:
Simplify how examples are numericalized

Remove unused options and code

gcampax committed:
Remove all the model parts related to the question

The question part of MQAN is always a fixed string and always
useless.

gcampax committed:
Remove revtok

gcampax committed:
Cleanup embedding code and command-line flags

Remove all flags that are always disabled (like the Transformer
layers on the decoder side) and set the dimension based on
the pretrained model size.

s-jse commented on genienlp/arguments.py:
```diff
@@ -99,9 +99,6 @@ def parse_argv(parser):
     parser.add_argument('--val_batch_size', nargs='+', default=[4000], type=int,
                         help='Number of tokens in each batch for validation, corresponding to tasks in --val_tasks')
     
-    parser.add_argument('--paired', action='store_true',
```
We don't actively use it but I think @Mehrad0711 wanted to do experiments with it in the future.

gcampax replied:
Yeah but the code for paired numericalization was gone before this PR.

Mehrad0711 replied:
We had some improvements using --paired training in SPL experiments; but it wasn't enough to warrant its inclusion.
I prefer to have it but if it complicates the API we can drop it for now. I will do more tests and if it's conclusive we can bring it back.

s-jse commented on genienlp/arguments.py:
```diff
@@ -139,63 +130,20 @@ def parse_argv(parser):
 
-    parser.add_argument("--almond_has_multiple_programs", action='store_true', help='Indicate if almond dataset has multiple programs for each sentence')
+    parser.add_argument('--model', type=str, choices=['BertLSTM', 'Bart'], default='BertLSTM', help='which model to import')
+    parser.add_argument('--pretrained_model', default=None,
```
It is not BertLSTM anymore then. TransformerLSTM maybe?

gcampax replied:
I'd like to keep BertLSTM because that's the name we used in the papers.

s-jse replied:
The SPL paper calls it XLM-R-LSTM, so we could opt for a more general name.

Mehrad0711 replied:
Yeah, I would also prefer the general term since the encoder can literally be any model from huggingface now that specific tokenizers are no longer needed.

Mehrad0711 commented on genienlp/arguments.py:
```diff
@@ -99,9 +99,6 @@ def parse_argv(parser):
-    parser.add_argument('--paired', action='store_true',
-                        help='Pair related examples before numericalizing the input (e.g. training with synthetic and paraphrase '
-                             'sentence pairs for almond task)')
     parser.add_argument('--max_pairs', type=int, default=1000000,
```
This should be omitted too with --paired gone.

Mehrad0711 commented on genienlp/util.py:
```diff
@@ -40,9 +40,7 @@
-from .data_utils.progbar import prange
 
 logger = logging.getLogger(__name__)
 
```
I assume some of the amond specific pre/ post processing functions (e.g. `multiwoz_specific_preprocess` ) will be gone too with this PR?

Mehrad0711 commented on genienlp/validate.py:
```diff
@@ -47,8 +47,7 @@ def generate_with_model(model, data_iterator, numericalizer, task, args, predict
+
     for batch in data_iterator:
         batch_size = len(batch.example_id)
         batch_prediction = [[] for _ in range(batch_size)] # a list where each element is a list of outputs for one input
```
if-else on line 67 should be omitted as well.

Mehrad0711 commented on tests/test.sh:
```diff
@@ -65,10 +55,10 @@ done
-      "--encoder_embeddings=bert-base-multilingual-uncased --decoder_embeddings= --trainable_decoder_embeddings=50 --seq2seq_encoder=Identity --dimension=768 --rnn_zero_state cls --almond_lang_as_question" ;
+      "--model BertLSTM --pretrained_model bert-base-multilingual-uncased --trainable_decoder_embeddings=50" \
+      "--model BertLSTM --pretrained_model bert-base-multilingual-uncased --trainable_decoder_embeddings=50 --sentence_batching --train_batch_tokens 4 --val_batch_size 4 --use_encoder_loss" \
+      "--model BertLSTM --pretrained_model bert-base-multilingual-uncased --trainable_decoder_embeddings=50 --sentence_batching --train_batch_tokens 4 --val_batch_size 4 --use_encoder_loss --paired" \
```
--paired test no longer needed.

Mehrad0711 commented on genienlp/models/base.py:
```diff
@@ -0,0 +1,72 @@
+            logger.info(f'Building vocabulary')
+            self.numericalizer.build_vocab(vocab_sets, tasks)
+
+    def add_new_vocab_from_data(self, tasks, resize_decoder=False):
```
Perhaps this method should be model specific (one for models with LSTM decoder and one for transformer seq2seq models)

gcampax replied:
There is nothing model specific to the interface: both models need it.

Mehrad0711 replied:
Not all GenieModels have self.bart and self.decoder

gcampax replied:
Yes I moved those bits to the subclass.

Mehrad0711 commented on genienlp/tasks/almond/__init__.py:
```diff
@@ -147,58 +144,19 @@ def _detokenize_cjk_chars(self, sentence):
 
-    def tokenize(self, sentence, field_name=None):
+    def preprocess_field(self, sentence, field_name=None):
+        if self.override_context and field_name == 'context':
```
We were doing "NFD" normalization in masked_tokenizers before to reduce number of OOVs in BERT. We should add it somewhere here.  
Also `tokenize_chinese_chars` and `strip_accents` for BertTokenizer need to be True unconditionally to mimic previous behavior.

gcampax replied:
Actually, we need to pass `do_basic_tokenize=False` to BertTokenizer otherwise ThingTalk symbols get all messed up, so those flags have no effect. We'll need to do things differently.

Mehrad0711 reviewed (commented):
Nice. This cleanup was much needed. A couple of comments for now:

s-jse commented on genienlp/util.py:
```diff
@@ -703,37 +633,16 @@ def load_config_json(args):
-                       'train_context_embeddings_after', 'train_question_embeddings_after'):
+            elif r == 'trainable_decoder_embedding':
                 setattr(args, r, 0)
             elif r == 'pretrain_mlm_probability':
```
`pretrain_mlm_probability` should be removed from here too.

s-jse commented on genienlp/util.py:
```diff
@@ -703,37 +633,16 @@ def load_config_json(args):
                 setattr(args, r, args.dimension)
             elif r == 'rnn_zero_state':
                 setattr(args, r, 'zero')
             elif r == 'use_pretrained_bert':
```
What does `use_pretrained_bert` do? Is it used anywhere?

gcampax replied:
If set to false, we load the architecture of the named bert (ie bert-base or bert-large) but then initialize it randomly. We meant to use it to see the effect of pretraining. I don't remember if we actually used it, or if it's implemented at all. Probably not.

s-jse commented on genienlp/data_utils/numericalizer.py:
```diff
@@ -0,0 +1,179 @@
+        batch: a List of List of integers
+        """
+        #TODO account for left padding models
+        return pad_sequence(batch, padding_value=self.pad_id, batch_first=True)
```
This unfortunately will not work with all HuggingFace models, and especially not if we want to have model-specific inputs (like BERT's `segment_id`). transformers' tokenizers don't have separate model-specific padding implementation, so this is probably the best we can do if we want to keep padding in `NumericalizedExamples.collate_batches()`.

gcampax replied:
Mh there is a Tokenizer.pad method: https://huggingface.co/transformers/internal/tokenization_utils.html#transformers.tokenization_utils_base.PreTrainedTokenizerBase.pad

s-jse replied:
I didn't see this! It is "internal" helper, but might be a good solution.

s-jse commented on genienlp/export.py:
```diff
@@ -51,14 +52,11 @@ def main(args):
-                                             args.decoder_embeddings,
-                                             args.max_generative_vocab,
-                                             logger)
+    config = AutoConfig.from_pretrained(args.pretrained_model, cache_dir=args.embeddings)
```
Do we actually use `export.py`?

gcampax replied:
Yes we use it in the Thingpedia training code (which is currently non-functional but we need in the future)

s-jse commented on genienlp/models/bart.py:
```diff
@@ -0,0 +1,99 @@
+        # TODO pretraining
+        if self.training:
+            batch = input[0]
+            pretraining = kwargs.pop("pretraining", None)
```
`pretraining` should be removed.

s-jse commented on genienlp/models/base.py:
```diff
@@ -0,0 +1,72 @@
+        self.numericalizer.grow_vocab(tasks)
+        if self.numericalizer.num_tokens > old_num_tokens:
+            logger.info(f'Vocabulary has expanded to {self.numericalizer.num_tokens} tokens')
+        self.bart.resize_token_embeddings(self.numericalizer.num_tokens)
```
`self.bart`?

s-jse commented on genienlp/models/common.py:
```diff
@@ -40,11 +40,6 @@
 from torch.nn.utils.rnn import pad_packed_sequence as unpack
 
 
-class EmbeddingOutput(NamedTuple):
```
`TransformerDecoder` and `mask_tokens()` in `common.py` seem to be unused as well.

s-jse reviewed (commented):
The changes make sense. I added a few small comments. Also, now that we are making these changes, we can also remove these last decaNLP names:
- Rename decaScore to genieScore.
- Rename `MQANDecoder` to `LSTMDecoder`.

And:
- Rename `IdentityEncoder` to `TransformerEncoder`

gcampax committed:
Fix tests, address review comments

gcampax committed:
Remove a lot of unused code

s-jse committed:
Log example lengths

gcampax committed:
Fix evaluation

gcampax committed:
tasks/almond: add additional preprocessing to the context

To make it closer to natural language

gcampax committed:
Add automatic preprocessing of special ThingTalk tokens

Automatically compute a replacement for all @ tokens, ^^ tokens,
and entity tokens, into natural-language-like words. The replacement
is based on the longest unambiguous suffix after splitting on
delimiters.

gcampax committed:
Add some more almond specific preprocessing to the thingtalk code

Replace mouthfuls QUOTED_STRING and GENERIC_ENTITY with STRING
and ENTITY

gcampax committed:
Remove bogus call to add_special_tokens

gcampax committed:
Fix numericalization

We need to use Huggingface's method of placing <s> and </s> or
we will confuse the model and train wrong.

gcampax committed:
Fix decoder start token in BART

gcampax committed:
Fix check

gcampax committed:
Remove very broken BART finetuning code

And rely on the correct label shifting done by the BART code

gcampax committed:
Remove BOS from the input to BART

Code analysis of fairseq and empirical observations suggest that
Huggingface is buggy in this respect.

gcampax committed:
Actually, only remove BOS from BART-Large

BART-Base definitely needs BOS

gcampax committed:
Add debugging around failures to compute token preprocessing

gcampax committed:
Restore BOS at the beginning of BART input sentences

Empirically, it works better

gcampax committed:
Merge remote-tracking branch 'origin/master' into wip/thingtalk2

gcampax committed:
tests: ensure we run T5 with --preprocess-special-tokens

T5 models don't know how to extend their vocabulary properly, so
they cannot deal with new special tokens and must preprocess all
special tokens into regular words.

gcampax committed:
numericalizer: ensure all needed special tokens are set

With TransformerLSTM, we need at least bos_token and eos_token
on the decoder side, so make sure they are set for all models,
in particular for BERT-style models that only have [CLS] and [SEP]

gcampax committed:
numericalizer: only use fast tokenizers that are known to work

Unfortunately, not all fast tokenizers work: some ignore
do_basic_tokenize, so they mess up the spacing/symbols in the input,
and some mess up the spacing around special tokens when reversing.

Add a list of models that are known to work with fast tokenizers.
Everything else is slow until tested and known to work.

gcampax committed:
Fix some LGTM warnings

s-jse commented on genienlp/data_utils/example.py:
```diff
@@ -27,217 +27,93 @@
 
     @staticmethod
-    def collate_batches(batches):
+    def collate_batches(batches : Iterable['NumericalizedExamples'], numericalizer, device):
```
In `Iterable['NumericalizedExamples']` do we need `''`? Which python version introduced this?

gcampax replied:
We need `''` _until_ Python 3.9, because the name `NumericalizedExamples` is not in scope at that point (due to Python's weird scoping rules for classes). All versions of Python with type annotations and the typing module support this (so IIRC 3.5+)
In Python 3.9, they introduced lazy resolution of type annotations so we would not need it.

gcampax committed:
Merge branch 'master' into wip/thingtalk2

s-jse reviewed (approved):
Looks good. `model_utils/parallel_utils.py` seems to have been unused for a while now.
On a separate note, does the code still work with multi-GPU training?
